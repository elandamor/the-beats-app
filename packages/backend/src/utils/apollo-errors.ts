import * as assert from "assert";
import ExtendableError from "extendable-error";

const isString = d => Object.prototype.toString.call(d) === "[object String]";
const isObject = d => Object.prototype.toString.call(d) === "[object Object]";
const isEmptyObject = d => Object.getOwnPropertyNames(d).length < 1;

export interface ErrorConfig {
  message: string;
  time_thrown?: string;
  data?: object;
  internalData?: object;
  options?: {
    showPath?: boolean;
    showLocations?: boolean;
  };
}

export interface ErrorInfo {
  code: string;
  field?: string;
  message: string;
  time_thrown: string;
  data?: object;
  path?: string;
  locations?: any;
}

export class ApolloError extends ExtendableError {
  code: string;
  field: string;
  message: string;
  time_thrown: string;
  data: object;
  internalData: object;
  path: any;
  locations: any;
  _showLocations: boolean = false;
  _showPath: boolean = false;

  // NOTE: The object passed to the Constructor is actually `ctorData`.
  //       We are binding the constructor to the code and config object
  //       for the first two parameters inside of `createError`
  constructor(code: string, config: ErrorConfig, ctorConfig: ErrorConfig) {
    super(
      (ctorConfig && ctorConfig.message) || (config && config.message) || ""
    );

    const t =
      (ctorConfig && ctorConfig.time_thrown) ||
      (config && config.time_thrown) ||
      new Date().toISOString();
    const m =
      (ctorConfig && ctorConfig.message) || (config && config.message) || "";
    const ctorData = (ctorConfig && ctorConfig.data) || {};
    const ctorInternalData = (ctorConfig && ctorConfig.internalData) || {};
    const configData = (config && config.data) || {};
    const configInternalData = (config && config.internalData) || {};
    const d = { ...this.data, ...configData, ...ctorData };
    const id = {
      ...this.internalData,
      ...configInternalData,
      ...ctorInternalData
    };
    const ctorOptions = (ctorConfig && ctorConfig.options) || {};
    const configOptions = (config && config.options) || {};
    const opts = { ...configOptions, ...ctorOptions };

    this.code = code;
    this.message = m;
    this.time_thrown = t;
    this.data = d;
    this.internalData = id;
    this._showLocations = !!opts.showLocations;
    this._showPath = !!opts.showPath;
  }

  serialize(): ErrorInfo {
    const {
      code,
      field,
      message,
      time_thrown,
      data,
      _showLocations,
      _showPath,
      path,
      locations
    } = this;

    const error: ErrorInfo = {
      code,
      ...(field && { field }),
      message,
      time_thrown,
      ...(!isEmptyObject(data) && { data }),
      path,
      locations
    };

    if (_showLocations) {
      error.locations = locations;
    }

    if (_showPath) {
      error.path = path;
    }

    return error;
  }
}

export const isInstance = e => e instanceof ApolloError;

export const createError = (code: string, config: ErrorConfig): ApolloError => {
  assert(
    isObject(config),
    "createError requires a config object as the second parameter"
  );
  assert(
    isString(config.message),
    'createError requires a "message" property on the config object passed as the second parameter'
  );
  // NOTE: The first two parameters give to the constructor will always be code and config
  //       Parameters passed to the constructor when `new` is invoked will be passed as
  //       subsequent parameters.
  return ApolloError.bind(null, code, config);
};

export const formatError = (error, returnNull = false): ErrorInfo => {
  const originalError = error ? error.originalError || error : null;

  if (!originalError) return returnNull ? null : error;

  const { code } = originalError;

  if (!code || !isInstance(originalError)) return returnNull ? null : error;

  const { time_thrown, data, _showLocations, _showPath } = originalError;
  const { message, locations, path } = error;

  if (message) {
    originalError.field =
      message.split(";").length > 1
        ? message
            .split(";")[1]
            .trim()
            .split(/\svalue./)[1]
        : undefined;
  }

  if (_showLocations) {
    originalError.locations = locations;
  }

  if (_showPath) {
    originalError.path = path;
  }

  return originalError.serialize();
};
