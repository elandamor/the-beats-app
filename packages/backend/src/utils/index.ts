import * as jwt from "jsonwebtoken";
import { Track } from "../generated/prisma-client";
import { APP_SECRET, JWT_EXPIRES_IN } from "../constants";
/**
 * getDuration
 * @param media - Track or Tracks
 */
export const getDuration = (media: Track | Track[]) => {
  const initialValue = 0;

  const computeDuration = (duration: string | number) => {
    // Convert duration to 'seconds' of type 'number' if input is of type 'string' with '(hh:)mm:ss' format
    return typeof duration === "string" ? timeToSeconds(duration) : duration;
  };

  if (!Array.isArray(media)) {
    return computeDuration(media.duration);
  }

  return media.reduce((previousTrack, currentTrack) => {
    const duration = computeDuration(currentTrack.duration);
    return previousTrack + duration;
  }, initialValue);
};

/**
 * Convert seconds(Int) to time(String)
 * @param secs
 */
export const secondsToTime = secs => {
  const hours = Math.floor(secs / 3600);
  const minutes = Math.floor((secs % 3600) / 60);
  const seconds = Math.floor((secs % 3600) % 60);

  return (
    (hours === 0
      ? ""
      : hours > 0 && hours.toString().length < 2
      ? "0" + hours + ":"
      : hours + ":") +
    (minutes.toString().length < 2 ? "0" + minutes : minutes) +
    ":" +
    (seconds.toString().length < 2 ? "0" + seconds : seconds)
  );
};

/**
 * Convert time(String) to seconds(Int)
 * @param arg - time as a string
 */
export const timeToSeconds = arg => {
  const time = arg.split(":");
  const seconds = parseInt(time[0], 10) * 60 + parseInt(time[1], 10);

  return seconds;
};

/**
 * Get userId of logged in account from context
 */
export const getUserId = request => {
  const authorization = request.headers.authorization;

  if (!authorization) {
    throw new Error("Authentication required!");
  }

  const token = authorization.replace("Bearer ", "");
  const decoded = jwt.verify(token, APP_SECRET);

  // @ts-ignore - (decoded) Property 'userId' does not exist on type 'string | object'
  return decoded.userId;
};

/**
 * Generates JWT token from user id
 */
export const generateToken = (userId: string) =>
  jwt.sign({ userId }, APP_SECRET, {
    expiresIn: JWT_EXPIRES_IN
  });
