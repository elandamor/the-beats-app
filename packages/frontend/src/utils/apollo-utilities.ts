import { ApolloLink, Operation, NextLink } from 'apollo-link';
import { onError } from 'apollo-link-error';

const formatError = (message: any, path?: any) => {
  if (path) {
    const headerCss = [
      'color: gray; font-weight: lighter', // title
      'color: #333; font-weight: 600', // message
      'color: gray; font-weight: lighter', // on
      'color: #333; font-weight: 600', // path
      'color: gray; font-weight: lighter', // request.
    ];

    const parts = ['%c message'];

    parts.push(`%c${message}`);
    parts.push('%con');
    parts.push(`%c${path.join('.')}`);
    parts.push('%crequest');

    return [parts.join(' '), ...headerCss];
  }

  const headerCss = [
    'color: gray; font-weight: lighter', // title
    'color: #333; font-weight: 600', // message
  ];

  const parts = ['%c message'];

  parts.push(`%c${message}`);

  return [parts.join(' '), ...headerCss];
};

const formatMessage = (operationType: any, operation: any, elapsed?: any) => {
  const headerCss = [
    'color: gray; font-weight: lighter', // title
    `color: ${operationType === 'query' ? '#02B875' : '#03A9F4'};`, // operationType
    'color: #000;', // operationName
  ];

  const parts = ['%c apollo'];

  parts.push(`%c${operationType}`);
  parts.push(`%c${operation.operationName}`);

  if (elapsed) {
    parts.push(`%c(in ${elapsed} ms)`);
    headerCss.push('color: gray; font-weight: lighter;'); // time
  }

  return [parts.join(' '), ...headerCss];
};

const errorLink = onError(({ graphQLErrors, networkError, operation }: any) => {
  if (graphQLErrors) {
    const errorType = 'graphQLError';
    const group = formatMessage(errorType, operation);

    console.groupCollapsed(...group);

    graphQLErrors.map(({ message, path }: any) => {
      const error = formatError(message, path);
      console.log(...error);
      return { message, path };
    });

    // @ts-ignore
    console.groupEnd(...group);
  }

  if (networkError) {
    const errorType = 'networkError';
    const group = formatMessage(errorType, operation);

    console.groupCollapsed(...group);

    const error = formatError(networkError.message);
    console.log(...error);

    // @ts-ignore
    console.groupEnd(...group);
  }
});

const loggerLink = new ApolloLink(
  (operation: Operation, forward?: NextLink): any => {
    const startTime = new Date().getTime();

    return (
      forward &&
      forward(operation).map((result) => {
        // @ts-ignore
        const operationType = operation.query.definitions[0].operation;
        const elapsed = new Date().getTime() - startTime;

        const group = formatMessage(operationType, operation, elapsed);
        console.groupCollapsed(...group);

        console.log('INIT', operation);
        console.log('RESULT', result);

        // @ts-ignore
        console.groupEnd(...group);

        return result;
      })
    );
  },
);

export { errorLink, loggerLink };
