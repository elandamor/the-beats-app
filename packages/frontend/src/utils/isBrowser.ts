import isNil from 'lodash/isNil';

const hasDocument = typeof document === 'object' && document !== null;
const hasWindow =
  typeof window === 'object' && window !== null && window.self === window;

const isBrowser: Function = () =>
  // @ts-ignore - Property 'override' does not exist on type 'Function'
  !isNil(isBrowser.override) ? isBrowser.override : hasDocument && hasWindow;

export default isBrowser;
