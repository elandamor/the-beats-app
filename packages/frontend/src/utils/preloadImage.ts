/**
 * Preload an image
 * @param {url} URL - The url of the image to preload
 */
const preloadImage = (url: string) => {
  return new Promise(function(resolve, reject) {
    const image = new Image();
    image.src = url;
    image.onload = resolve;
    image.onerror = reject;
  });
};

export default preloadImage;
