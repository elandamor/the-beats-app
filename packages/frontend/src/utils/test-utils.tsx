import { render } from '@testing-library/react';
import { AppProvider } from '@app/contexts';

// TODO: Make customRender type-safe.
const customRender = (ui: any, options?: any) =>
  render(ui, { wrapper: AppProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
