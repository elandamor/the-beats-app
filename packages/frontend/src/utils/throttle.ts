/**
 * A throttled function that only invokes `func` at most once per every `threshold` milliseconds.
 * @param func The function to throttle.
 * @param threshold The number of milliseconds to throttle invocations to.
 * @param scope  Where variables, functions, and objects are accessible in your code during runtime
 */

export default function throttle(
  func: Function,
  threshold: number = 250,
  scope?: any,
) {
  /**
   * @var last The last time `func` was executed.
   * @var timeoutID ID of the timeout you wish to clear, as returned by setTimeout().
   */
  let last: number, timeoutID: number | NodeJS.Timeout;

  return function(this: any) {
    let context = scope || this;
    let now = Date.now(),
      args = arguments;

    if (last && now < last + threshold) {
      if (typeof timeoutID === 'number') {
        clearTimeout(timeoutID);
      }

      timeoutID = setTimeout(function() {
        last = now;
        func.apply(context, args);
      }, threshold);
    } else {
      last = now;
      func.apply(context, args);
    }
  };
}
