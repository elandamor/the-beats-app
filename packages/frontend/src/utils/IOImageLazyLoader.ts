import preloadImage from './preloadImage';

export function updateSessionStorage(src: string) {
  const cachedImages = JSON.parse(
    window.sessionStorage.getItem('__TRA_IMG__') || '{}',
  );
  cachedImages[src] = +new Date();
  window.sessionStorage.setItem('__TRA_IMG__', JSON.stringify(cachedImages));
}

/**
 * Check browser support for IntersectionObserver(IO)
 * @return {boolean} true or false - True if browser supports IO
 */
function SUPPORTS_INTERSECTION_OBSERVER() {
  return 'IntersectionObserver' in window;
}
/**
 * Get css-class to set on image once it has been handled by IO
 * @return {string} css-class - The css class attached to a handled image
 */
function HANDLED_CLASS() {
  return 'lazy-image-handled';
}
/**
 * Get css-class to set on image once it has been loaded by preloader
 * @return {string} css-class - The css class attached to a loaded image
 */
function LOADED_CLASS() {
  return 'lazy-image-loaded';
}

/**
 * Add the respective image to the DOM
 */
function _applyImage(img: any, src: string) {
  const el = img;

  el.classList.add(LOADED_CLASS());

  if (!el) {
    return;
  }
  // If target element is an image ...
  if (el.tagName.toLowerCase() === 'img') {
    // ... set src attribute of element ...
    el.src = src;
  } else {
    // ... otherwise set image source as background of element.
    el.style.backgroundImage = `url(${src})`;
  }

  updateSessionStorage(src);
}
/**
 * Preload the respective image
 * @param {element} img - The image to be loaded
 * @return {callback} _applyImage - A method that adds the image to the DOM
 */
function _preloadImage(image: any) {
  // Get the source for the image.
  const src = image.dataset.src;
  // Do nothing if image source is missing.
  if (!src) {
    return;
  }

  return preloadImage(src).then(() => _applyImage(image, src));
}

/**
 * Class to lazyload images
 */
export default function IOLazyImageLoader(image: any) {
  // Load image immediately if IO is not supported
  if (SUPPORTS_INTERSECTION_OBSERVER) {
    _preloadImage(image);
    return;
  }
  // Setup IO
  const io = new IntersectionObserver((entries, observer) => {
    entries.forEach((entry) => {
      if (entry.isIntersecting) {
        entry.target.classList.add(HANDLED_CLASS());

        _preloadImage(entry.target);

        observer.unobserve(entry.target);
      }
    });
  });
  // ... observe image
  io.observe(image);
}
