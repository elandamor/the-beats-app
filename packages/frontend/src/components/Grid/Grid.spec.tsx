// Grid.spec.tsx
import * as React from 'react';
import { render } from '@app/utils/test-utils';

import Grid from './index';

describe('Grid', () => {
  it('should display a grid', () => {
    const { container } = render(
      <Grid columns={2}>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </Grid>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
