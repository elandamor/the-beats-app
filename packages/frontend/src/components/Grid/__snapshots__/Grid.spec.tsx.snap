// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should display a grid 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(2,1fr);
  grid-gap: 24px;
}

<div
  class="c-grid c0"
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;
